File: leet-code/2020/0012-integer-to-roman.cpp
// https://leetcode.com/problems/integer-to-roman/

class Solution {
public:
    string intToRoman(int num) {
        std::string roman;
        for (int i(0); num; num /= 10, ++i) {
            switch (num%10) {
                case 4:
                    roman = convertNumberFour(i) + roman;
                    break;
                case 5:
                    roman = convertNumberFive(i) + roman;
                    break;
                case 9:
                    roman = convertNumberNine(i) + roman;
                    break;
                case 0:
                    break;
                default:
                    roman = convertRegularNumbers(num%10, i) + roman;
            }
            
        }
        return roman;
    }
    
private:
    std::string convertNumberFour(int iteration) {
        switch (iteration) {
            case 0:
                return "IV";
            case 1:
                return "XL";
            case 2:
                return "CD";
            default:
                throw std::runtime_error("impossible");
        }
    }

    std::string convertNumberNine(int iteration) {
        switch (iteration) {
            case 0:
                return "IX";
            case 1:
                return "XC";
            case 2:
                return "CM";
            default:
                throw std::runtime_error("impossible");
        }
    }

    std::string convertNumberFive(int iteration) {
        switch (iteration) {
            case 0:
                return "V";
            case 1:
                return "L";
            case 2:
                return "D";
            default:
                throw std::runtime_error("impossible");
        }
    }
    
    std::string convertRegularNumbers(int number, int iteration) {
        std::vector<char> romanChars{'I', 'X', 'C', 'M'};
        std::string result;
        if (number > 4) {
            result = convertNumberFive(iteration);
        }
        return result + std::string(number%5, romanChars.at(iteration));
    }
};

File: leet-code/2021/0012-integer-to-roman.rs
https://leetcode.com/problems/integer-to-roman/

impl Solution {
    pub fn int_to_roman(num: i32) -> String {
        let mut num = num;
        let mut result = String::new();
        let mut roman_letters: &[char] = &['Z', 'Y', 'M', 'D', 'C', 'L', 'X', 'V', 'I'];
        let mut divisor = 1000;

        const Tens: usize = 0;
        const Fives: usize = 1;
        const Units: usize = 2;

        while num > 0 {
            result.push_str(&match num / divisor {
                n @ 0..=3 => format!("{}",   roman_letters[Units]).repeat(n as usize),
                4         => format!("{}{}", roman_letters[Units], roman_letters[Fives]),
                n @ 5..=8 => format!("{}{}", roman_letters[Fives], format!("{}", roman_letters[Units]).repeat((n%5) as usize)),
                9         => format!("{}{}", roman_letters[Units], roman_letters[Tens]),
                _ => panic!("not possible")
            });
            num %= divisor;
            divisor /= 10;
            roman_letters = &roman_letters[2..];
        }
        result
    }
}

impl LessHeapSolution {
    pub fn int_to_roman(num: i32) -> String {
        let mut num = num;
        let mut result = String::new();
        let mut roman_letters: &[char] = &['Z', 'Y', 'M', 'D', 'C', 'L', 'X', 'V', 'I'];
        let mut divisor = 1000;

        const Tens: usize = 0;
        const Fives: usize = 1;
        const Units: usize = 2;

        while num > 0 {
            match num / divisor {
                n @ 0..=3 => {
                    for _ in 0..n { result.push(roman_letters[Units]); }
                },
                4 => {
                    result.push(roman_letters[Units]);
                    result.push(roman_letters[Fives]);
                },
                n @ 5..=8 => {
                    result.push(roman_letters[Fives]);
                    for _ in 0..(n%5) { result.push(roman_letters[Units]); }
                },
                9 => {
                    result.push(roman_letters[Units]);
                    result.push(roman_letters[Tens]);
                },
                _ => panic!("not possible")
            }
            num %= divisor;
            divisor /= 10;
            roman_letters = &roman_letters[2..];
        }
        result
    }
}

impl AlternativeSolution {
    pub fn int_to_roman(num: i32) -> String {
        let mut num = num;
        let mut result = String::new();
        let mut roman_letters = "ZYMDCLXVI";
        let mut divisor = 1000;

        const Tens: usize = 0;
        const Fives: usize = 1;
        const Units: usize = 2;

        while num > 0 {
            match num / divisor {
                n @ 0..=3 => {
                    for _ in 0..n { result.push(Self::get(roman_letters, Units)); }
                },
                4 => {
                    result.push(Self::get(roman_letters, Units));
                    result.push(Self::get(roman_letters, Fives));
                },
                n @ 5..=8 => {
                    result.push(Self::get(roman_letters, Fives));
                    for _ in 0..(n%5) { result.push(Self::get(roman_letters, Units)); }
                },
                9 => {
                    result.push(Self::get(roman_letters, Units));
                    result.push(Self::get(roman_letters, Tens));
                },
                _ => panic!("not possible")
            }
            num %= divisor;
            divisor /= 10;
            roman_letters = &roman_letters[2..];
        }
        result
    }

    fn get(roman_letters: &str, n: usize) -> char {
        roman_letters.chars().nth(n).unwrap()
    }
}
